Percentile/Ranks should return null instead of NaN when empty

polyfractal opened the pull request:
The other metric aggregations (min/max/etc) return `null` as their XContent value and string when nothing was computed (due to empty/missing fields).  Percentiles and Percentile Ranks, however, return `NaN `which is inconsistent and confusing for the user.  This fixes the inconsistency by making the aggs return `null`.  This applies to both the numeric value and the "as string" value.  

Note: like the metric aggs, this does not change the value if fetched directly from the percentiles object, which will return as `NaN`/`"NaN"`. This only changes the XContent output.

I looked through all the other metric aggs and they appear to return null (or 0.0, in the case of cardinality/value_count/sum).  So percentiles were the only outliers.

This is sorta a bwc break, but could also be seen as a bugfix.  I'm not sure what we want to do with regards to backporting.

Closes #29066

elasticmachine commented:
Pinging @elastic/es-search-aggs

cbuescher commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/percentiles/hdr/InternalHDRPercentilesRanksTests.java:
```diff
@@ -103,4 +113,85 @@ protected InternalHDRPercentileRanks mutateInstance(InternalHDRPercentileRanks i
         return new InternalHDRPercentileRanks(name, percents, state, keyed, formatter, pipelineAggregators, metaData);
     }
+
+    public void testEmptyRanksXContent() throws IOException {
```
I'm just looking how similar the xContent output of InternalHDRPercentilesRanksTests and InternalTDigestPercentilesRanksTest, maybe these two test could be pushed up one level to InternalPercentilesRanksTestCase by calling the sub-tests createTestInstance() method with the appropriate values? I haven't really checked if the outputs are exactly the same, maybe I'm missing something, but it would be great to reduce the number of rather identical test cases.
Maybe pushing all four cases up to AbstractPercentilesTestCase would work as well? Not sure though.

polyfractal commented:
Ran into a bit of a snag. Good call on adding the "empty" case to the general xcontent roundtrip tests @cbuescher.  Exposed some broken behavior, where the xcontent was serialized with `null` but deserialized into `NaN`, causing problems.

The issue is that Percentiles, Ranks, Stats and ExtendedStats all extend `NumericMetricAggregator.MultiValue`, which defines `double metric(String name, long owningBucketOrdinal)`.  While investigating if we can change that return signature to `Double`, I found this in the Stats agg:

```
if (valuesSource == null || owningBucketOrd >= counts.size()) {
            switch(InternalStats.Metrics.resolve(name)) {
                case count: return 0;
                case sum: return 0;
                case min: return Double.POSITIVE_INFINITY;
                case max: return Double.NEGATIVE_INFINITY;
                case avg: return Double.NaN;
                default:
                    throw new IllegalArgumentException("Unknown value [" + name + "] in common stats aggregation");
            }
        }
```

Extended stats is similar, using a mix of `NaN` and `+/- Inf` to signal "missing".  I feel like we should:

1. Change the return signature of the method to `Double` so that we can return null here too
2. Include Stats/ExtendedStats in the refactoring to use `null` instead of the mixture of `NaN` and `+/- Inf`

Thoughts @cbuescher @colings86 ?

colings86 commented:
@polyfractal This is tricky because the reason Stats/ExtendedStats outputs those values is to match the outputs of the individual aggs they are combining in those cases. For example, the `min` aggregation will return `Double.POSITIVE_INFINITY` in the empty case. This consistency between the individual aggs and the combined Stats/ExtendedStats aggs is important. So with what you propose we would have to change the min/max/etc. aggs too. This is starting to feel like a much bigger breaking change than the initial idea. 

I'm not against it but we need to be more careful as the impact of the breaking change increases. If we go down this route then we might need to think about migration and bwc since we don't want to surprise users too much with the break /cc @clintongormley

colings86 commented:
Additionally, outputting `null` might have an adverse reaction to the pipeline aggregations which I think will determine that the bucket_path is wrong rather than the value is just missing from the bucket if null is returned. We should make sure we test this

polyfractal commented:
> This is tricky because the reason Stats/ExtendedStats outputs those values is to match the outputs of the individual aggs they are combining in those cases. For example, the min aggregation will return Double.POSITIVE_INFINITY in the empty case. 

I went back through everything and think I understand how it works now.  It's a bit more weird, as it turns out.  Here's the situation for the metrics:

- InternalMin [emits a `null`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/metrics/min/InternalMin.java#L86) if the value is infinite† when serializing to XContent output
- ParsedMin also [emits a `null`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/metrics/min/ParsedMin.java#L44) like InternalMin when serializing to XContent
- When deserializing from XContent, Min converts `null` values [back into `Inf`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/metrics/min/ParsedMin.java#L54)
- However, both [InternalMin](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/metrics/min/InternalMin.java#L70) and [ParsedMin](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/metrics/min/ParsedMin.java#L32) return the actual double value from the getter.  So Java TC users, and pipeline aggs, will get our internal placeholder value which is different values from REST


This seems to hold for all metrics, stats, etc.  The test failures I was encountering was due to not adjusting the ParsedPercentiles `doXContentBody` to do the same trick as InternalPercentiles (emitting null if there were no percentiles).  So if I follow the same pattern, xcontent serializes into `null`, but deserializes back into `NaN` internally.

It then follows the same properties as the other aggs: getters and pipeline aggs show our internal placeholder value (`Inf`, `NaN`, etc) while XContent shows `null`.  This still feels messy but it should keeps the breaking changes to a minimum.  I'll polish it up and push a commit soon.

†Interestingly, Min doesn't check _which_ infinity is present, which means a legitimate `-Inf` value would be treated as if it were `null`.

polyfractal commented:
Jenkins, run gradle build tests

polyfractal commented:
Hmm, there seems to be something consistently broken with the empty xcontent tests, as it keeps failing on CI.  But I just can't reproduce locally, even with 100,000 iterations on the same or random seed.

```
./gradlew :server:test -Dtests.seed=2C59C2697A193E1 -Dtests.class=org.elasticsearch.search.aggregations.metrics.percentiles.tdigest.InternalTDigestPercentilesRanksTests -Dtests.method="testEmptyRanksXContent" -Dtests.security.manager=true -Dtests.locale=bg-BG -Dtests.timezone=America/North_Dakota/New_Salem
15:13:21 FAILURE 0.03s J1 | InternalTDigestPercentilesRanksTests.testEmptyRanksXContent <<< FAILURES!
15:13:21    > Throwable #1: java.lang.AssertionError: 
15:13:21    > Expected: "NaN"
15:13:21    >      but: was "�"
15:13:21    > 	at __randomizedtesting.SeedInfo.seed([2C59C2697A193E1:3C1FA52E15FFD98]:0)
15:13:21    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
```
Will keep poking at it.

polyfractal commented:
Ok, I've tested this extensively locally and could never reproduce... and the CI is now passing after merging master.  So I'm tentatively claiming it was a weird CI issue.

@cbuescher mind taking another look?  I think I addressed the issues you raised.

cbuescher commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/percentiles/InternalPercentilesRanksTestCase.java:
```diff
@@ -39,4 +49,52 @@ protected final void assertFromXContent(T aggregation, ParsedAggregation parsedA
         assertTrue(parsedClass != null && parsedClass.isInstance(parsedAggregation));
     }
+
+    public void testEmptyRanksXContent() throws IOException {
```
You mentioned earlier you cannot push this test up into AbstractPercentilesTestCase because of some subtle difference, but I cannot spot it. Do you remember what it was? Otherwise I'd give it another try to push it up.

polyfractal replied:
It's super tiny:  Percentiles uses `percent() / percentAsString()` while PercentileRanks uses `percentile() / percentileAsString()`.

I could collapse them into a single test and then do an `instanceOf` or `getType()` and switch on that if you think it'd be cleaner.  Less test code duplication, but a bit more fragile.

cbuescher replied:
Ah, I see it now. What about pulling the test up and just doing the two lines of assertions that are different in their own little helper method that you overwrite differently in both cases? I'm usually also not a fan of doing so much code acrobatics in tests but in this case I think the gain in non-duplicated lines of code would justify it. I don't think its super important though, thanks for pointing out the difference.

polyfractal replied:
++ this cleaned up nicely.  Thanks for the suggestion!

polyfractal commented:
Tests cleaned up and nicely de-duplicated, and added a note to the breaking changes doc

polyfractal commented:
Thanks @cbuescher.  The mystery failure is back, so there must be something here that I'm missing.  Going to go back over the PR and see if I missed something.

cbuescher commented:
@polyfractal does it look the same as the one mentioned above? What I find strange it that is doesn't reproduce in that case. That probably needs some investigation too.

polyfractal commented:
Yeah, it's another one of these unprintable characters.  Makes me think there's a serialization issue and something isn't being written/read correctly.

```
REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=8E1D067ECFCE26E4 -Dtests.class=org.elasticsearch.search.aggregations.metrics.percentiles.hdr.InternalHDRPercentilesRanksTests -Dtests.method="testEmptyRanksXContent" -Dtests.security.manager=true -Dtests.locale=zh-TW -Dtests.timezone=America/Panama
15:39:09 FAILURE 0.02s J2 | InternalHDRPercentilesRanksTests.testEmptyRanksXContent <<< FAILURES!
15:39:09    > Throwable #1: java.lang.AssertionError: 
15:39:09    > Expected: "NaN"
15:39:09    >      but: was "�"
15:39:09    > 	at __randomizedtesting.SeedInfo.seed([8E1D067ECFCE26E4:8F19600AB930489D]:0)
15:39:09    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
15:39:09    > 	at org.elasticsearch.search.aggregations.metrics.percentiles.InternalPercentilesRanksTestCase.assertPercentile(InternalPercentilesRanksTestCase.java:48)
15:39:09    > 	at org.elasticsearch.search.aggregations.metrics.percentiles.AbstractPercentilesTestCase.testEmptyRanksXContent(AbstractPercentilesTestCase.java:111)
```

jasontedor commented:
Note that we run with `9-:-Djava.locale.providers=COMPAT` in the default `jvm.options`. I had been thinking last week about the fact that we do not run our tests with the options that we ship with in the default `jvm.options` file. This feels like a build infrastructure problem that we should address.

polyfractal commented:
Jenkins, run gradle build tests

polyfractal commented:
Alrighty, looks like we're back to a green build.  @cbuescher would you mind taking a look at the most recent commit to see if you approve?  :heart:

cbuescher commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalNumericMetricsAggregation.java:
```diff
@@ -79,7 +79,12 @@ protected MultiValue(StreamInput in) throws IOException {
 
         public String valueAsString(String name) {
-            return format.format(value(name)).toString();
+            // Explicitly check for NaN, since it formats to "�" or "NaN" depending on JDK version
```
Is this only the case for certain locales? I would bne suprised if some JDKs would return a weird UTF8 character in all cases. To make this comment more readable it would probably also make sense to put in the bad utf8 value as octal or hex codepoint and to clarify under which circumstances this happens.

polyfractal replied:
I'm not actually sure how this behaves across Locales, but I don't think it matters for us.  We seem to always initialize the Decimal `DocValueFormat` [with `Locale.Root`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/DocValueFormat.java#L366) which I believe uses the JRE's default symbol table.  

So for JDK8 the root locale will use `JRELocaleProviderAdapter` to get the symbols, which loads `sun.text.resources.FormatData`, and you can see the [`NaN` symbol is `\uFFFD`](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/text/resources/FormatData.java#l269)

For JDK 9+, the root locale will use `CLDRLocaleProviderAdapter`, which loads `sun.text.resources.cldr.FormatData`.  And in that resource file you can see the `NaN` symbol is `"NaN"` (Can't find a link to the code, but you can see it in your IDE).

++ to making the comment more descriptive.  I'll try to distill this thread into a sane comment, and probably leave a reference to the comments here in case anyone wants to see more info.

polyfractal replied:
As an aside, I really wonder why Oracle thought � would be a good default representation of "NaN"... :(

polyfractal replied:
Hm, I don't like how this was implemented, looking at it.  Going to move it over to the DocValueFormat itself, so that it only applies to the Decimal formatter when looking at doubles... otherwise it'll be checked against all formatters (geo, IP, etc).  Harmless I think, but no need.

cbuescher commented on server/src/main/java/org/elasticsearch/search/DocValueFormat.java:
```diff
@@ -394,6 +394,22 @@ public String format(long value) {
+             *
+             * Since the character � isn't very useful, and makes the output change depending on JDK version,
+             * we manually check to see if the value is NaN and return the string directly.
+             */
```
+1
Great comment, my future self will be glad its here ;-)

polyfractal replied:
:)  Me too!
